docs/specifications/tests/RequirementsAndNormalizationOfDomainNames.md

```diff
@@ -0,0 +1,516 @@
+
+1. a valid ASCII domain name, or
+2. a valid IDN name (Internationalized Domain Name) as of
+   [IDNA2008][RFC 5890#1.1].
+
+The process defined in this specification will normalize *Domain Name* and output
+a normalized form to be used by all Zonemaster test cases. The objectives of the
+normalization are
+
+1. Remove leading and trailing white space characters.
+2. Convert other dot characters to regular dot (or "FULL STOP").
+3. Create legal IDNA 2008 U-labels from convenient alternative forms.
+4. Create consistent representation of the same zone name.
+
+The result of the normalization can be a new form of *Domain Name* to be used
+by the tests in test cases, the normalized form. If the normalized form is
+neither a valid ASCII domain name nor a valid IDN name, then *Domain Name*
+cannot be used for Zonemaster testing.
+
+If the outcome (see [Outcome(s)](#Outcomes)) is not "fail" then *Domain Name* in
+normalized form is returned to be used as input value for Zonemaster test cases.
+
+See the details in the [Detailed requirements] section below.
+
+
+### References
+
+The following references are consulted for this specification:
+
+* [RFC 1034]
+* [RFC 1035]
+* [RFC 1123]
+* [RFC 2317]
+* [RFC 2782]
+* [RFC 5890]
+* [RFC 5891]
+* [RFC 5895]
+* [Unicode TR 46]
+
+
+## Scope
+
+This specification only tests and creates a normalized form of the domain name
+(zone name or name server name).
+
+In this specification, ASCII is identical to the first 128 characters in
+[Unicode] (0000..007F).
+
+[RFC 1123][RFC 1123#2.1], section 2.1, specifies that a domain name label
+may not start or end with a HYPHEN-MINUS ("-"), only digit or letter. This
+restriction on HYPHEN-MINUS is disregarded in this specification and is assumed
+to be handled in test case [Syntax02].
+
+The use of the SOLIDUS ("/") and the LOW LINE ("_") in domain name is discussed
+in the section "[ASCII domain name](#ASCII-domain-name)" below. Any restrictions
+on where in the domain name or label those could or should be used are
+disregarded in this specification, and are assumed to be handled in test cases
+[Syntax01] and [Syntax02].
+
+
+## Inputs
+
+* "Domain Name" - The domain name to be tested and normalized according to this
+  specification. It must be a non-empty string of [Unicode] characters.
+
+## Summary
+
+In the specification there are six scenarios that will result in the domain name
+not being usable, i.e. it cannot be used for Zonemaster testing. Each scenario
+is here listed with a message tag, level (always CRITICAL in this specification),
+suitable argument to be used in the same descriptive message and a message that
+can be returned to the user.
+
+Message Tag                    | Level    | Arguments | Message ID for message tag
+:------------------------------|:---------|:----------|:----------------------------------------------------------
+CAPITAL_I_DOT_ABOVE_UNSUPPORTED| CRITICAL |           | LATIN CAPITAL LETTER I WITH DOT ABOVE cannot be downcased.
```

Comment:
Wouldn't it be better if we use a more generic error message taking the code point or character as argument? Like
`AMBIGUOUS_DOWN_CASING | CRITICAL | codepointname | {codepointname} cannot be downcased.`

I feel that the current one is very specific, but if there is no other case than this one I don't mind that much.

Reply:
Sure. Also works. Updated.